enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md

```diff
@@ -0,0 +1,253 @@
+
+Each index adds load and overhead (e.g. mapping, metadata) to the Elasticsearch cluster that needs to be tracked beyond the actual data.   Elasticsearch has recommendations for maximum shard size and the number of shards per node per allocated gigabyte of heap.  Cluster logging typically exceeds these recommendations for any OpenShift clusters that has significant log traffic.
+
+
+### Goals
+
+
+The goals of this proposals are:
+* Utilize a data design that aligns data schema with Elasticsearch's recommendations.
+* Expose data management policy as API in the `cluster-logging-operator` and `elasticsearch-operator` in support of Cluster Logging's mission to gather container and infrastructure logs
+ 
+
+### Non-Goals
+
+This change will not:
+* Provide a general data management policy API that fully exposes Elasticsearch's rollover API
+
+## Proposal
+
+This proposal introduces two specific changes to to achieve its goals:
+
+* Co-located data in a few opinionated indices
+* Index management using [rollover index API](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/indices-rollover-index.html)
+
+Logs of a given type (e.g. app container, infra container, node) are separated by index.  The Cluster Logging collector writes logs to a well-known alias established by the `cluster-logging-operator`.  The ClusterLogging CR instance specifies the management policy for each log type index.  This controls: 
+* maximum number of logs per index (e.g. 30G) 
+* the maximum age (e.g. 7 days).  
+
+This policy is passed to the Elasticsearch CR for the `elasticsearch-operator` to manage the rollover policy.
+
+
+### Implementation Details
+
+#### Assumptions
+
+* ClusterLogging will expose the minimal needed set of Elasticsearch rollover policy management API in order to achieve the previously described goals
+* ClusterLogging will manage rollover as Elasticsearch index management is either restricted by Elastic licensing or not available in the opensource version of OpenDistro
+* Security will be addressed by using the OpenDistro security plugin and document level security (DLS). Details TBD.
+
+#### Data Model
+Logs of a given type are co-located to the following indices:
+
+| Log Type | Read Alias | Write Alias | Initial Index |
+|---------|-----------|---------------|---------------|
+| Node (`logs.infra`)|infra, infra.journal|infra.journal-write|infra.journal-00001|
+| Container Infra (`logs.infra`)|infra, infra.container|infra.container-write|infra.container-000001|
+| Container Application (`logs.app`)|app.logs|app.container-write|app.container-000001|
+
+**Note:** Log types are further defined in [LogForwarding](./cluster-logging-log-forwarding.md).
+
+#### ClusterLogging API
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogging"
+metadata:
+  name: "instance"
+spec:
+  logStore:
+    retentionPolicy: 
+      logs.app:
+        maxAge: 7d
+      logs.infra
+        maxAge: 7d
```

Comment:
it seems awkward to me that we have retention policies at both the elastic + clusterlogging config levels..what happens when they conflict w/ each other?

Reply:
> it seems awkward to me that we have retention policies at both the elastic + clusterlogging config levels..what happens when they conflict w/ each other?

by both do you mean we express something on CL and we express something on 'Elastic'.  As mentioned here: https://github.com/openshift/enhancements/pull/108/files#r346466953 CLO design goal has been to be a higher level abstraction where Elasticsearch allows more 'knob' tuning.  All of the "Elasticsearch" settings would be generated by CLO so its fundamentally not "public" to consumers of cluster logging

Reply:
yeah that's what i meant, but if you're saying that the user/admin will not be allowed to edit the EO config settings (or they will be stomped by the CLO if they do) then that's fine.

Reply:
to close the loop on this, EO is the final source of truth when it comes to policing the actual Elasticsearch cluster.
CLO (when it's CR is in a `managed` state) will end up enforcing that the elasticsearch CR spec matches what it has in its own CR.